docs/tuning.md

```diff
@@ -264,6 +264,13 @@ parent RDD's number of partitions. You can pass the level of parallelism as a se
 or set the config property `spark.default.parallelism` to change the default.
 In general, we recommend 2-3 tasks per CPU core in your cluster.
 
+Sometimes you may also need to increase directory listing parallelism when job input has large number of directories,
+otherwise the process could take a very long time, especially when against object store like S3.
+If your job works on RDD with Hadoop input formats (e.g., via `SparkContext#sequenceFile`), the parallelism is
+controlled via `spark.hadoop.mapreduce.input.fileinputformat.list-status.num-threads` (default is 1). For other
```

Comment:
This seems having a limitation that multiple threads cannot be used with non thread-safe path filter?

https://hadoop.apache.org/docs/r2.7.2/hadoop-mapreduce-client/hadoop-mapreduce-client-core/mapred-default.xml

> The number of threads to use to list and fetch block locations for the specified input paths. Note: multiple threads should not be used if a custom non thread-safe path filter is used. 

Should we also mention it together?

Reply:
I think this is pretty rare and also that most users won't probably be exposed to this (they mostly interact with RDDs and file formats, I think). Plus this is a Hadoop configuration (recognizable from the name) so they can also find the other doc online.

Reply:
Maybe, shall we add a hyperlink on `spark.hadoop.mapreduce.input.fileinputformat.list-status.num-threads` to the hadoop configuration, https://hadoop.apache.org/docs/r3.2.0/hadoop-mapreduce-client/hadoop-mapreduce-client-core/mapred-default.xml? Then, we can lead them safely if they didn't take a look at that. It should be Hadoop 3.2.0 link on `master` branch.

Reply:
SGTM

Reply:
Sounds good. I'll add a link to the master branch.